### DISCUSSION TITLE
Websocket Contract first API

### DISCUSSION THREAD

**Discussion by naleeha** [finos/vuu#1449] [Ideas]
### Considerations
- Auditability - can we see the message between front end and server
   - In the Chrome dev tools
   - In logs/audit trail  
- API contract testing
- Managing changes - e.g. API versioning
- Code gen to Scala & Typescript
- Document generation

### Potential solutions

**gRPC** 

- Protocol Buffer as IDL (Interface Description Language) - with validation
- Code gen to Scala & Typescript
- API contract testing - using Buf Lint test and the gRPC mock 

https://medium.com/adidoescode/defining-the-api-contract-a1e0c96cedd2
https://blog.rockthejvm.com/grpc-in-scala-with-fs2-scalapb/#3-the-proto-file
https://medium.com/cloud-native-daily/building-high-performance-microservices-with-node-js-grpc-and-typescript-ddef5e0bdb95

**Avro**
https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/serdes-avro.html

**Comment by chrisjstevo** [comment]
One thing to consider here is whether we view the typescript and the scala code as a single system or two separate systems.

Interface versioning within a single system is a lot less valuable. It is the internal message protocol. Versioning where we might join different versions at runtime is valuable, but hasn't been an assumption so far.

The codegen piece might be more interesting, although we use Jackson on the server side to effective generate the encoders & decoders dynamically.
